feat(task-worker): Add taskworker pull

enochtangg opened the pull request:
This PR is responsible for adding the taskworker pull which is responsible for fetching tasks from a task buffer (inflight actiivation store) and executing them in a multiprocess pool. After tasks are executed, the worker will update the task status appropriately in the task queue.

Review notes / Changes:
* Default `expires` and `processing_deadline_duration` are defined on the namespace level. If these fields are not provided in the task, the namespace defaults will be used.
* The `process_deadline_duration` is defined on the task activation. When the arroyo consumer stores an inflight activation into the buffer, it will use this value to set the `processing_deadline`.
* The `TaskClient` is a dummy class created for testing. Once we unblock GRPC, we can properly implement this.

TODOs:
- [x] Merge https://github.com/getsentry/sentry/pull/80072 and use the new TaskRegistry with namespace routing. This currently uses the old version.
- [x] Merge updated protobuf definition: https://github.com/getsentry/sentry-protos/pull/59
- [x] Fix multiprocess unable to import appropriate modules. When it tries to, we see a `Apps aren't loaded yet.` error. Django has not finished setting up (e.g. loading installed apps, etc.) before we are trying to access settings and importing module. This is causes the following unit tests to fail: `test_taskworker_process_task_complete`, `test_taskworker_process_task_retry`, and `test_taskworker_process_task_timeout_failure`. 
- [x] Add any missing tests.

enochtangg committed:
wip

enochtangg committed:
clean up

enochtangg committed:
unit tests

enochtangg committed:
clean up

enochtangg commented on src/sentry/conf/server.py:
```diff
@@ -1298,6 +1298,8 @@ def SOCIAL_AUTH_DEFAULT_USERNAME() -> str:
 CELERY_TASK_SOFT_TIME_LIMIT = int(timedelta(hours=3).total_seconds())
 CELERY_TASK_TIME_LIMIT = int(timedelta(hours=3, seconds=15).total_seconds())
 
+TASKWORKER_IMPORTS = ("tests.sentry.taskworker.test_taskworker",)
```
TODO: Patch this setting in the test that runs it

markstory commented on src/sentry/taskworker/registry.py:
```diff
@@ -85,3 +98,39 @@ def send_task(self, activation: TaskActivation) -> None:
 
 
 # TODO(taskworker) Import TaskRegistrys
+class TaskRegistry:
```
This is part of #80072

markstory commented on src/sentry/taskworker/task.py:
```diff
@@ -69,6 +73,7 @@ def create_activation(self, *args: P.args, **kwargs: P.kwargs) -> TaskActivation
             parameters=orjson.dumps({"args": args, "kwargs": kwargs}).decode("utf8"),
             retry_state=self._create_retry_state(),
             received_at=received_at,
+            processing_deadline_duration=self._processing_deadline_duration,
```
We'll need to add `expires` here too.

markstory commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+
+class TaskWorker:
+    """
+    A TaskWorker fetches tasks from a task queue (inflight actiivation store) and handles executing them.
```
```suggestion
    A TaskWorker fetches tasks from a task queue (inflight activation store) and handles executing them.
```

markstory commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+        self._execution_count = 0
+        self._worker_id = uuid4().hex
+        self._namespace: TaskNamespace | None = None
+        self._build_pool()
```
We might want to defer the pool construction until `start()`. That might help with django setup errors too.

markstory committed:
Update sentry-protos

john-z-yang commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+        self._build_pool()
+
+    def _build_pool(self) -> None:
+        self._pool = Pool(processes=2)
```
Why does this process pool has 2 processes? Isn't a single one sufficient?

markstory replied:
Updated to be 1 now.

john-z-yang commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+                "taskworker.task.change_status",
+                extra={"task": activation.taskname, "state": next_state},
+            )
+            response = task_client.set_task_status(
+                task_id=activation.id,
+                task_status=next_state,
+            )
```
Why do we get another task back again immediately? I remember us talking about this but I don't quite remember why. Looking at the actual code, it feels like this is a little too hard to understand and maintain.

john-z-yang replied:
bit of a nitpick, but I think if we want to do both in 1 grpc call, could we separate the execution of the task and sending result back to the server into 2 functions?

markstory replied:
Getting the 'next' task back in the response of a status change lets the worker only make a single RPC call to advance, instead of needing to make one call for status update and another to get the next task.

markstory committed:
Allow timedelta for expires and processing_deadline_duration

Remove idempotent as it will come back later with a different name and
wasn't implemented before.

john-z-yang commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+
+                next_task, processing_deadline = self.process_task(task, processing_deadline)
+
+                if max_task_count is not None and max_task_count <= self._execution_count:
+                    self.exitcode = 1
+                    raise WorkerCompleteException("Max task exeuction count reached")
+
+        except KeyboardInterrupt:
+            self.exitcode = 1
+        except WorkerCompleteException as err:
+            self.exitcode = 0
+            logger.warning(str(err))
```
I feel like it might be easier to maintain if we just return the exit code here, instead of creating a custom exception and catching it
```python
if max_task_count is not None and max_task_count <= self._execution_count:
    logger.warning("Max task execution count reached")
    return 1
```

john-z-yang commented on src/sentry/taskworker/taskworker.py:
```diff
@@ -0,0 +1,190 @@
+            logger.info("taskworker.task_execution_timeout")
+            # When a task times out we kill the entire pool. This is necessary because
+            # stdlib multiprocessing.Pool doesn't expose ways to terminate individual tasks.
+            self._pool.terminate()
```
Let's put this in build pool, so we don't forget to call this when we build

markstory committed:
Simplify tests and start cleaning up the worker logic.

markstory committed:
Merge branch 'master' into taskworker-worker-pull

markstory committed:
return exitcode instead of storing it as an attribute.

markstory committed:
Expand coverage for start()

markstory committed:
Streamline worker and get it running more efficiently

markstory committed:
Fix mypy

markstory committed:
Patch settings for tests.

markstory committed:
Merge branch 'master' into taskworker-worker-pull

markstory committed:
Align create_namespace with TaskNamespace

markstory committed:
Remove taskworker-dlq as it won't be used by sentry

markstory committed:
Merge branch 'master' into taskworker-worker-pull

markstory committed:
Add test coverage for ctrl-c and unknown tasks being handled gracefully.

markstory committed:
Make default value consistent

evanh commented on src/sentry/taskworker/worker.py:
```diff
@@ -0,0 +1,208 @@
+        Run the worker main loop
+
+        Once started a Worker will loop until it is killed, or
+        completes its max_task_count when it shuts down.
```
Why do workers have a max task count?

evanh commented on src/sentry/taskworker/worker.py:
```diff
@@ -0,0 +1,208 @@
+        except Exception as err:
+            if namespace.get(activation.taskname).should_retry(activation.retry_state, err):
+                logger.info("taskworker.task.retry", extra={"task": activation.taskname})
+                next_state = TASK_ACTIVATION_STATUS_RETRY
```
It feels odd that the worker is deciding if a task should be retried or not. Shouldn't the worker just report a failure, and the consumer upkeep would decide to retry it or not?

markstory replied:
Retry policies include allow/block lists for exception types that have to be evaluated in the worker. The workers make a suggestion to the consumer that this task should be retried *if possible*. The consumer will make a final decision based on the number of retries a task has used and how many are available.
